File: leet-code/2. Add Two Numbers/AddTwoNumbers2.java
import java.math.BigInteger;

public class AddTwoNumbers2 {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode r1 = reverseLinkedList(l1);
        String strNum1 = "";
        while (r1 != null) {
            strNum1 += Integer.toString(r1.val);
            r1 = r1.next;
        }

        ListNode r2 = reverseLinkedList(l2);
        String strNum2 = "";
        while (r2 != null) {
            strNum2 += Integer.toString(r2.val);
            r2 = r2.next;
        }

        BigInteger b1 = new BigInteger(strNum1);
        BigInteger b2 = new BigInteger(strNum2);
        BigInteger bSums = b1.add(b2);
        ListNode b = bigIntegerToListNode(bSums);
        return reverseLinkedList(b);
    }

    private ListNode reverseLinkedList(ListNode node) {
        return reverse(node, null);
    }

    private ListNode reverse(ListNode cur, ListNode prev) {
        if (cur == null) {
            return prev;
        }

        ListNode next = cur.next;
        cur.next = prev;
        return reverse(next, cur);
    }

    private ListNode bigIntegerToListNode(BigInteger b) {
        char[] charArr = b.toString().toCharArray();
        ListNode head = new ListNode(Character.getNumericValue(charArr[0]));
        ListNode node = head;
        for (int i = 1; i < charArr.length; i++) {
            node.next = new ListNode(Character.getNumericValue(charArr[i]));
            node = node.next;
        }
        return head;
    }
}

File: leet-code/2. Add Two Numbers/AddTwoNumbers2.kt
package com.example

class AddTwoNumbers2 {
    fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
        var list1 = l1
        var list2 = l2
        val root = ListNode(0)
        // 올림수
        var carry = 0
        var cur = root
        while (list1 != null || list2 != null || carry != 0) {
            // 현재합
            var sum = 0
            if (list1 != null) {
                sum += list1.`val`
                list1 = list1.next
            }

            if (list2 != null) {
                sum += list2.`val`
                list2 = list2.next
            }

            sum += carry
            carry = sum / 10
            val remainder = sum % 10
            cur.next = ListNode(remainder)
            cur = cur.next!!
        }
        return root.next
    }
}
